Micronaut data fails when trying to convert a single row to Flow

### Expected Behavior

When using a return type of Flow<MyEntity> and when the given query only returns one result I expect the conversion to work to return a Flow that emits a single item.

### Actual Behaviour

When the query returns a single item. Micronaut fails to convert that entity to a Flow.

### Steps To Reproduce

1. Create a `CoroutineCrudRepository<MyEntity, Long>` repository and entity `MyEntity`.
2. Run the `.findAll()` query after inserting a single row.
3. The query will fail with an error message below.

```
io.micronaut.core.convert.exceptions.ConversionErrorException: Cannot convert type [class com.example.MyEntity] to target type: interface kotlinx.coroutines.flow.Flow. Considering defining a TypeConverter bean to handle this case.
	at app//io.micronaut.core.convert.ConversionService.lambda$newConversionError$2(ConversionService.java:182)
	at java.base@17.0.1/java.util.Optional.orElseGet(Optional.java:364)
	at app//io.micronaut.core.convert.ConversionService.newConversionError(ConversionService.java:182)
	at app//io.micronaut.core.convert.ConversionService.lambda$convertRequired$0(ConversionService.java:177)
	at java.base@17.0.1/java.util.Optional.orElseThrow(Optional.java:403)
	at app//io.micronaut.core.convert.ConversionService.convertRequired(ConversionService.java:177)
	at app//io.micronaut.core.convert.ConversionService.convertRequired(ConversionService.java:159)
	at app//io.micronaut.data.runtime.intercept.AbstractQueryInterceptor.convertOne(AbstractQueryInterceptor.java:213)
	at app//io.micronaut.data.runtime.intercept.AbstractQueryInterceptor.convertOne(AbstractQueryInterceptor.java:202)
	at app//io.micronaut.data.runtime.intercept.async.AbstractConvertCompletionStageInterceptor.lambda$intercept$1(AbstractConvertCompletionStageInterceptor.java:61)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base@17.0.1/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at app//reactor.core.publisher.MonoToCompletableFuture.onNext(MonoToCompletableFuture.java:64)
	at app//reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)
	at app//reactor.core.publisher.MonoUsingWhen$MonoUsingWhenSubscriber.deferredComplete(MonoUsingWhen.java:269)
	at app//reactor.core.publisher.FluxUsingWhen$CommitInner.onComplete(FluxUsingWhen.java:528)
	at app//reactor.core.publisher.MonoFlatMap$FlatMapMain.secondComplete(MonoFlatMap.java:250)
	at app//reactor.core.publisher.MonoFlatMap$FlatMapInner.onComplete(MonoFlatMap.java:324)
	at app//reactor.core.publisher.MonoFlatMap$FlatMapMain.secondComplete(MonoFlatMap.java:250)
	at app//reactor.core.publisher.MonoFlatMap$FlatMapInner.onComplete(MonoFlatMap.java:324)
	at app//reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:155)
	at app//reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74)
	at app//reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2571)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2367)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2241)
	at app//reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
	at app//reactor.core.publisher.Mono.subscribe(Mono.java:4512)
	at app//reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onComplete(FluxSwitchIfEmpty.java:82)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:2231)
	at app//reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144)
	at app//reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:2231)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onComplete(MonoIgnoreThen.java:210)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onComplete(MonoIgnoreThen.java:210)
	at app//reactor.pool.SimpleDequePool.maybeRecycleAndDrain(SimpleDequePool.java:537)
	at app//reactor.pool.SimpleDequePool$QueuePoolRecyclerInner.onComplete(SimpleDequePool.java:767)
	at app//reactor.core.publisher.Operators.complete(Operators.java:137)
	at app//reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:46)
	at app//reactor.core.publisher.Mono.subscribe(Mono.java:4512)
	at app//reactor.pool.SimpleDequePool$QueuePoolRecyclerMono.subscribe(SimpleDequePool.java:879)
	at app//reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:241)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onComplete(MonoIgnoreThen.java:204)
	at app//reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260)
	at app//reactor.core.publisher.Operators.complete(Operators.java:137)
	at app//reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:46)
	at app//reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:76)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:264)
	at app//reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51)
	at app//reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:241)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onComplete(MonoIgnoreThen.java:204)
	at app//reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onComplete(MonoIgnoreElements.java:89)
	at app//reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onComplete(FluxHandleFuseable.java:239)
	at app//reactor.core.publisher.MonoSupplier$MonoSupplierSubscription.request(MonoSupplier.java:148)
	at app//reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.request(FluxHandleFuseable.java:260)
	at app//reactor.core.publisher.MonoIgnoreElements$IgnoreElementsSubscriber.onSubscribe(MonoIgnoreElements.java:72)
	at app//reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onSubscribe(FluxHandleFuseable.java:164)
	at app//reactor.core.publisher.MonoSupplier.subscribe(MonoSupplier.java:48)
	at app//reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:76)
	at app//reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.subscribeNext(MonoIgnoreThen.java:264)
	at app//reactor.core.publisher.MonoIgnoreThen.subscribe(MonoIgnoreThen.java:51)
	at app//reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:76)
	at app//reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:53)
	at app//reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:76)
	at app//reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:165)
	at app//reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74)
	at app//reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2571)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2367)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2241)
	at app//reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
	at app//reactor.core.publisher.Mono.subscribe(Mono.java:4512)
	at app//reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onComplete(FluxSwitchIfEmpty.java:82)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:2231)
	at app//reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144)
	at app//reactor.core.publisher.Operators.complete(Operators.java:137)
	at app//reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:46)
	at app//reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:76)
	at app//reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:165)
	at app//reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onNext(FluxSwitchIfEmpty.java:74)
	at app//reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2571)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2367)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:2241)
	at app//reactor.core.publisher.MonoJust.subscribe(MonoJust.java:55)
	at app//reactor.core.publisher.Mono.subscribe(Mono.java:4512)
	at app//reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onComplete(FluxSwitchIfEmpty.java:82)
	at app//reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:2231)
	at app//reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:144)
	at app//reactor.core.publisher.Operators.complete(Operators.java:137)
	at app//reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:46)
	at app//reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:76)
	at app//reactor.core.publisher.FluxUsingWhen$UsingWhenSubscriber.onComplete(FluxUsingWhen.java:385)
	at app//reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126)
	at app//reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:850)
	at app//reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:612)
	at app//reactor.core.publisher.FluxFlatMap$FlatMapMain.innerComplete(FluxFlatMap.java:898)
	at app//reactor.core.publisher.FluxFlatMap$FlatMapInner.onComplete(FluxFlatMap.java:1001)
	at app//reactor.core.publisher.FluxHandle$HandleSubscriber.onComplete(FluxHandle.java:223)
	at app//reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onComplete(MonoFlatMapMany.java:260)
	at app//reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onComplete(FluxHandleFuseable.java:239)
	at app//reactor.core.publisher.FluxFilterFuseable$FilterFuseableConditionalSubscriber.onComplete(FluxFilterFuseable.java:391)
	at app//reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onComplete(FluxContextWrite.java:126)
	at app//reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onComplete(FluxPeekFuseable.java:940)
	at app//reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onComplete(FluxPeekFuseable.java:940)
	at app//io.r2dbc.postgresql.util.FluxDiscardOnCancel$FluxDiscardOnCancelSubscriber.onComplete(FluxDiscardOnCancel.java:104)
	at app//reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onComplete(FluxDoFinally.java:128)
	at app//reactor.core.publisher.FluxHandle$HandleSubscriber.onComplete(FluxHandle.java:223)
	at app//reactor.core.publisher.FluxCreate$BaseSink.complete(FluxCreate.java:465)
	at app//reactor.core.publisher.FluxCreate$BufferAsyncSink.drain(FluxCreate.java:871)
	at app//reactor.core.publisher.FluxCreate$BufferAsyncSink.complete(FluxCreate.java:819)
	at app//reactor.core.publisher.FluxCreate$SerializedFluxSink.drainLoop(FluxCreate.java:249)
	at app//reactor.core.publisher.FluxCreate$SerializedFluxSink.drain(FluxCreate.java:215)
	at app//reactor.core.publisher.FluxCreate$SerializedFluxSink.complete(FluxCreate.java:206)
	at app//io.r2dbc.postgresql.client.ReactorNettyClient$Conversation.complete(ReactorNettyClient.java:668)
	at app//io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.emit(ReactorNettyClient.java:934)
	at app//io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:810)
	at app//io.r2dbc.postgresql.client.ReactorNettyClient$BackendMessageSubscriber.onNext(ReactorNettyClient.java:716)
	at app//reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:129)
	at app//reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onNext(FluxPeekFuseable.java:854)
	at app//reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:224)
	at app//reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:224)
	at app//reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:294)
	at app//reactor.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:403)
	at app//reactor.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:426)
	at app//reactor.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:114)
	at app//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at app//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at app//io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at app//io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at app//io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at app//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at app//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at app//io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at app//io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at app//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at app//io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at app//io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at app//io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at app//io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at app//io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at app//io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at app//io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at app//io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at app//io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@17.0.1/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: Cannot convert type [class com.example.MyEntity] to target type: interface kotlinx.coroutines.flow.Flow. Considering defining a TypeConverter bean to handle this case.
	... 147 more
```

### Environment Information

MacOS 14.4

openjdk 17.0.1 2021-10-19 LTS
OpenJDK Runtime Environment Zulu17.30+15-CA (build 17.0.1+12-LTS)
OpenJDK 64-Bit Server VM Zulu17.30+15-CA (build 17.0.1+12-LTS, mixed mode, sharing)

### Example Application

_No response_

### Version

4.3.7

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/DataIntroductionAdvice.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/DataIntroductionAdvice.java
@@ -99,11 +99,11 @@ public final class DataIntroductionAdvice implements MethodInterceptor<Object, O
                 if (throwable == null) {
                     Class<Object> target = context.getReturnType().asArgument().getType();
                     if (value == null) {
-                        value = new NullValue();
+                        value = conversionService.convert(new NullValue(), target).orElse(value);
+                    } else {
+                        value = conversionService.convert(value, target).orElse(value);
                     }
-                    completableFuture.complete(
-                        conversionService.convert(value, target).orElse(value)
-                    );
+                    completableFuture.complete(value);
                 } else {
                     Throwable finalThrowable = throwable;
                     if (finalThrowable instanceof CompletionException) {

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/DataIntroductionAdvice.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/DataIntroductionAdvice.java
@@ -28,6 +28,7 @@ import io.micronaut.data.annotation.Repository;
 import io.micronaut.data.intercept.DataInterceptor;
 import io.micronaut.data.intercept.RepositoryMethodKey;
 import io.micronaut.data.runtime.convert.DataConversionService;
+import io.micronaut.data.runtime.support.NullValue;
 import io.micronaut.inject.InjectionPoint;
 import jakarta.inject.Inject;
 
@@ -96,7 +97,13 @@ public final class DataIntroductionAdvice implements MethodInterceptor<Object, O
         completionStage.whenComplete((value, throwable) -> {
             try (PropagatedContext.Scope ignore = propagatedContext.propagate()) {
                 if (throwable == null) {
-                    completableFuture.complete(value);
+                    Class<Object> target = context.getReturnType().asArgument().getType();
+                    if (value == null) {
+                        value = new NullValue();
+                    }
+                    completableFuture.complete(
+                        conversionService.convert(value, target).orElse(value)
+                    );
                 } else {
                     Throwable finalThrowable = throwable;
                     if (finalThrowable instanceof CompletionException) {

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/async/AbstractAsyncInterceptor.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/async/AbstractAsyncInterceptor.java
@@ -19,6 +19,7 @@ import io.micronaut.aop.MethodInvocationContext;
 import io.micronaut.core.annotation.NonNull;
 import io.micronaut.core.annotation.Nullable;
 import io.micronaut.core.type.Argument;
+import io.micronaut.core.type.ReturnType;
 import io.micronaut.data.exceptions.DataAccessException;
 import io.micronaut.data.operations.RepositoryOperations;
 import io.micronaut.data.operations.async.AsyncCapableRepository;
@@ -62,10 +63,13 @@ public abstract class AbstractAsyncInterceptor<T, R> extends AbstractQueryInterc
     }
 
     protected final Argument<?> findReturnType(MethodInvocationContext<?, ?> context, Argument<?> defaultArg) {
+        ReturnType<?> returnType = context.getReturnType();
         if (context.isSuspend()) {
-            return context.getReturnType().asArgument();
+            if (!returnType.getType().getName().equals("kotlinx.coroutines.flow.Flow")) {
+                return returnType.asArgument();
+            }
         }
-        return context.getReturnType().asArgument().getFirstTypeVariable().orElse(defaultArg);
+        return returnType.asArgument().getFirstTypeVariable().orElse(defaultArg);
     }
 
     /**

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/support/NullValue.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/support/NullValue.java
@@ -1,0 +1,28 @@
+/*
+ * Copyright 2017-2025 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.runtime.support;
+
+import io.micronaut.core.annotation.Internal;
+
+/**
+ * The value that represents a null for the conversion.
+ *
+ * @author Denis Stepanov
+ * @since 4.11
+ */
+@Internal
+public record NullValue() {
+}

--- a/data-runtime/src/main/kotlin/io/micronaut/data/runtime/support/convert/KotlinFlow.kt
+++ b/data-runtime/src/main/kotlin/io/micronaut/data/runtime/support/convert/KotlinFlow.kt
@@ -1,0 +1,84 @@
+/*
+ * Copyright 2017-2025 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.runtime.support.convert
+
+import io.micronaut.context.annotation.Prototype
+import io.micronaut.context.annotation.Requires
+import io.micronaut.core.annotation.Internal
+import io.micronaut.core.convert.MutableConversionService
+import io.micronaut.core.convert.TypeConverterRegistrar
+import io.micronaut.data.runtime.support.NullValue
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.future.await
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+
+/**
+ * The Kotlin Flow converters.
+ *
+ * @author Denis Stepanov
+ * @since 4.11
+ */
+@Requires(classes = [Flow::class])
+@Prototype
+@Internal
+internal class KotlinFlow : TypeConverterRegistrar {
+    override fun register(conversionService: MutableConversionService) {
+        conversionService.addConverter(
+            CompletableFuture::class.java,
+            Flow::class.java
+        ) { completableFuture: CompletableFuture<*> ->
+            flow {
+                val value = completableFuture.await()
+                emit(
+                    value
+                )
+            }
+        }
+        conversionService.addConverter(
+            CompletionStage::class.java,
+            Flow::class.java
+        ) { completableFuture: CompletionStage<*> ->
+            flow {
+                val value = completableFuture.await()
+                emit(
+                    value
+                )
+            }
+        }
+        conversionService.addConverter(
+            Object::class.java,
+            Flow::class.java
+        ) { obj: Any ->
+            flow {
+                emit(
+                    obj
+                )
+            }
+        }
+        conversionService.addConverter(
+            NullValue::class.java,
+            Flow::class.java
+        ) { _: Any ->
+            flow {
+                emit(
+                    null
+                )
+            }
+        }
+    }
+}

--- a/doc-examples/jdbc-example-kotlin/build.gradle
+++ b/doc-examples/jdbc-example-kotlin/build.gradle
@@ -34,4 +34,5 @@ dependencies {
     runtimeOnly mnSql.h2
 
     testImplementation mnTest.micronaut.test.kotest5
+    testImplementation 'org.jetbrains.kotlin:kotlin-test'
 }

--- a/doc-examples/jdbc-example-kotlin/src/main/kotlin/example/TestEntity.kt
+++ b/doc-examples/jdbc-example-kotlin/src/main/kotlin/example/TestEntity.kt
@@ -1,0 +1,21 @@
+package example
+
+import io.micronaut.core.annotation.Introspected
+import io.micronaut.data.annotation.DateCreated
+import io.micronaut.data.annotation.Id
+import io.micronaut.data.annotation.MappedEntity
+import java.time.Instant
+import java.util.*
+
+@MappedEntity("test_table")
+@Introspected
+data class TestEntity(
+    @field:Id
+    val id: UUID,
+    val destination: String,
+    val scheduledTime: Instant,
+    val payload: String,
+    val attempts: Int,
+    @DateCreated
+    val created: Instant?
+)

--- a/doc-examples/jdbc-example-kotlin/src/main/kotlin/example/TestRepository.kt
+++ b/doc-examples/jdbc-example-kotlin/src/main/kotlin/example/TestRepository.kt
@@ -1,0 +1,21 @@
+package example
+
+
+import io.micronaut.data.jdbc.annotation.JdbcRepository
+import io.micronaut.data.model.query.builder.sql.Dialect
+import io.micronaut.data.repository.kotlin.CoroutineCrudRepository
+import kotlinx.coroutines.flow.Flow
+import java.time.Instant
+import java.util.*
+
+
+@JdbcRepository(dialect = Dialect.H2)
+abstract class TestRepository: CoroutineCrudRepository<TestEntity, UUID> {
+    abstract suspend fun findByScheduledTimeBeforeAndAttemptsLessThanEquals(scheduledTime: Instant, attempts: Int): Flow<TestEntity>
+
+    /**
+     * Showcase that making this return a List instead of Flow works... idk the implications of this. Does it internally
+     * call Flow.toList()?
+     */
+    abstract suspend fun findByScheduledTimeBefore(scheduledTime: Instant): List<TestEntity>
+}

--- a/doc-examples/jdbc-example-kotlin/src/test/kotlin/example/KotlinFlowConversionRepositoryTest.kt
+++ b/doc-examples/jdbc-example-kotlin/src/test/kotlin/example/KotlinFlowConversionRepositoryTest.kt
@@ -1,0 +1,82 @@
+package example
+
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.Test
+import java.time.Instant
+import java.util.*
+import kotlin.test.assertEquals
+import kotlin.test.assertIs
+import kotlin.test.assertNotNull
+
+@MicronautTest
+class KotlinFlowConversionRepositoryTest(
+    private val testRepository: TestRepository
+) {
+
+    @Test
+    fun testEmptyTable() {
+        runTest {
+            val results = testRepository.findAll()
+            assertIs<Flow<TestEntity>>(results)
+            val entity = testRepository.findByScheduledTimeBeforeAndAttemptsLessThanEquals(Instant.now(), 5)
+            assertIs<Flow<TestEntity>>(entity)
+            assertEquals(0, results.toList().size)
+        }
+    }
+
+    @Test
+    fun testCannotConvert() {
+        runTest {
+            testRepository.save(TestEntity(
+                id = UUID.randomUUID(),
+                destination = "test",
+                scheduledTime = Instant.now().minusSeconds(10),
+                payload = "test",
+                attempts = 0,
+                created = Instant.now()
+            ))
+            var results = testRepository.findAll()
+            assertIs<Flow<TestEntity>>(results)
+            assertEquals(1, results.toList().size)
+            // Throws: Cannot convert type [class com.example.TestEntity] to target type: interface kotlinx.coroutines.flow.Flow. Considering defining a TypeConverter bean to handle this case.
+            results = testRepository.findByScheduledTimeBeforeAndAttemptsLessThanEquals(Instant.now(), 5)
+            assertIs<Flow<TestEntity>>(results)
+            assertEquals(1, results.toList().size)
+        }
+    }
+
+    @Test
+    fun testEmptyTableAsList() {
+        runTest {
+            val results = testRepository.findByScheduledTimeBefore(Instant.now())
+            assertNotNull(results)
+            assertIs<List<TestEntity>>(results)
+            assertEquals(0, results.size)
+        }
+    }
+
+    @Test
+    fun testWithListReturnType() {
+        runTest {
+            testRepository.save(TestEntity(
+                id = UUID.randomUUID(),
+                destination = "test",
+                scheduledTime = Instant.now().minusSeconds(10),
+                payload = "test",
+                attempts = 0,
+                created = Instant.now()
+            ))
+            val results = testRepository.findAll()
+            assertIs<Flow<TestEntity>>(results)
+            assertEquals(1, results.toList().size)
+            // Throws: Cannot convert type [class com.example.TestEntity] to target type: interface kotlinx.coroutines.flow.Flow. Considering defining a TypeConverter bean to handle this case.
+            val listResults = testRepository.findByScheduledTimeBefore(Instant.now())
+            assertIs<List<TestEntity>>(listResults)
+            assertEquals(1, listResults.size)
+        }
+    }
+
+}

